rllib/algorithms/dreamerv3/dreamerv3_rl_module.py

```diff
@@ -0,0 +1,140 @@
+
+
+@ExperimentalAPI
+class DreamerV3RLModule(RLModule, abc.ABC):
+    def setup(self):
+        # Gather model-relevant settings.
+        B = 1
+        T = self.config.model_config_dict["batch_length_T"]
+        horizon_H = self.config.model_config_dict["horizon_H"]
+        gamma = self.config.model_config_dict["gamma"]
+        symlog_obs = do_symlog_obs(
+            self.config.observation_space,
+            self.config.model_config_dict.get("symlog_obs", "auto"),
+        )
+        model_dimension = self.config.model_config_dict["model_dimension"]
+
+        # Build encoder and decoder from catalog.
+        catalog = self.config.get_catalog()
+        self.encoder = catalog.build_encoder(framework=self.framework)
+        self.decoder = catalog.build_decoder(framework=self.framework)
+
+        # Build the world model (containing encoder and decoder).
+        self.world_model = WorldModel(
+            model_dimension=model_dimension,
+            action_space=self.config.action_space,
+            batch_length_T=T,
+            # num_gru_units=self.model_config.num_gru_units,
+            encoder=self.encoder,
+            decoder=self.decoder,
+            symlog_obs=symlog_obs,
+        )
+        self.actor = ActorNetwork(
+            action_space=self.config.action_space,
+            model_dimension=model_dimension,
+        )
+        self.critic = CriticNetwork(
+            model_dimension=model_dimension,
+        )
+        # Build the final dreamer model (containing the world model).
+        self.dreamer_model = DreamerModel(
+            model_dimension=self.config.model_config_dict["model_dimension"],
+            action_space=self.config.action_space,
+            world_model=self.world_model,
+            actor=self.actor,
+            critic=self.critic,
+            # use_curiosity=use_curiosity,
+            # intrinsic_rewards_scale=intrinsic_rewards_scale,
+            batch_size_B=self.config.model_config_dict["batch_size_B"],
+            batch_length_T=T,
+            horizon_H=horizon_H,
+        )
+        self.action_dist_cls = catalog.get_action_dist_cls(framework=self.framework)
+
+        # Perform a test `call()` to force building the dreamer model's variables.
+        test_obs = np.tile(
+            np.expand_dims(self.config.observation_space.sample(), (0, 1)),
+            reps=(B, T, 1),
+        )
+        test_actions = np.tile(
+            np.expand_dims(
+                one_hot(
+                    self.config.action_space.sample(), depth=self.config.action_space.n
+                ),
+                (0, 1),
+            ),
+            reps=(B, T, 1),
+        )
+        self.dreamer_model(
+            inputs=test_obs,
+            actions=test_actions.astype(np.float32),
+            is_first=np.ones((B, T), np.float32),
+            start_is_terminated_BxT=np.zeros((B * T,), np.float32),
+            horizon_H=horizon_H,
+            gamma=gamma,
+        )
```

Comment:
wait, why is this needed?

Reply:
To define all variables in the module. If I don't make this call, module.trainable_variables will be empty and I won't be able to print out a summary of the model, which is super useful to see what we have actually built.

Reply:
An alternative to making this call would be to provide input shape information to each of the components, which, to be honest, is actually more work than performing this single call with proper test data. This seems to be a tf only problem as for torch, you would have to provide input shapes anyways when you build your nn.Modules.
However, if you don't make the call, the tf.Model is not considered "built" and we won't be able to do `summary()`, even if the variables are known. I agree, we should not print out the summary in general. Maybe just on the driver and only in debug logging mode.